### PR TITLE
[Dynamo x FSDP][3/x] TypedStorage and storage_offset

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106738
* #106906
* #106890
* __->__ #106888
* #106886
* #106884

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305 @ipiszy

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -88,6 +88,8 @@ def __init__(
         stride=None,
         is_contiguous=None,
         specialized_value=None,
+        _typed_storage=None,
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:91]: ...what exactly is this? This seems very questionable, because it's not a primitive type like size/stride

**Comment by ezyang** [comment]
Needs tests.

File: torch/_dynamo/variables/tensor.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:91]: This is the tensor's _typed_storage `torch.storage.TypedStorage` - do you want me to write a first class citizen handing for this that does not go through this init?

One thing we do here is just store the original _typed_storage, and then defer to a lambda variable below to make a TypedStorageVariable - if you like, we can try a different approach, like fakifying it eagerly, making TypedStorageVariable eagerly, and turning ops like `_resize_` below into ops.

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -1039,3 +1072,47 @@ def call_function(
...
+                unimplemented("Dynamic sizing in typed storage - NYI")
+            if isinstance(self.value._size(), int):
+                return ConstantVariable(self.value._size())
+            sizes = [ConstantVariable(x) for x in self.value._size()]
+            return SizeVariable(sizes)
+        if name == "device":
+            return ConstantVariable(self.value.device)
+        if name == "_resize_":
+            assert len(args) == 1
+            self.value._resize_(args[0].value)
```
Comment (voznesenskym) [inline torch/_dynamo/variables/tensor.py:1116]: should this be an op??

File: torch/_dynamo/variables/tensor.py [hunk 3]
```diff
@@ -1039,3 +1072,47 @@ def call_function(
             )
 
         return super().call_function(tx, args, kwargs)
+
+
+class TypedStorageVariable(VariableTracker):
+    def __init__(self, value, **kwargs):
+        self.value = value
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:1079]: What exactly is value here?

File: torch/_dynamo/variables/tensor.py [hunk 4]
```diff
@@ -1039,3 +1072,47 @@ def call_function(
...
+
+    def reconstruct(self, codegen):
+        return super().reconstruct(codegen)
+
+    def as_proxy(self):
+        return self.proxy
+
+    def var_getattr(self, tx, name: str) -> VariableTracker:
+        if name == "_size":
+            return variables.LambdaVariable(
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:1090]: This isn't right. If you getattr a method you get a bound method, this is not a lambda and they behave differently in different situations.

File: torch/_dynamo/variables/tensor.py [hunk 5]
```diff
@@ -1039,3 +1072,47 @@ def call_function(
...
+
+    def call_method(
+        self,
+        tx,
+        name,
+        args: "List[VariableTracker]",
+        kwargs: "Dict[str, VariableTracker]",
+    ) -> "VariableTracker":
+        if name == "_data_ptr":
+            return ConstantVariable(self.value._data_ptr())
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:1103]: This is exceedingly unlikely to be right, as you are now burning in pointer constants into the traced Dynamo graph. But typically, I would expect storages to have "tensor" nature and be varying across runs.

Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:1103]: Or if value is a fake storage, I would expect this to always be a null pointer, which would also be wrong to burn in as a constant variable!

File: torch/_dynamo/variables/tensor.py (hunk 4 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:1090]: This is the dynamo pattern for doing this - it doesn't create a lambda, its a lambda variable which, when invoked, calls method. (see the method below)

File: torch/_dynamo/variables/tensor.py (hunk 5 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:1103]: Agreed, I wonder if we should put storage objects in the graph...

File: torch/_dynamo/variables/tensor.py [hunk 6]
```diff
@@ -124,6 +128,14 @@ def check_type(ty):
         else:
             return check_type(tensor_type)
 
+    def call_hasattr(self, tx, name: str) -> "VariableTracker":
+        options = VariableTracker.propagate(self)
+        val = self.as_proxy().node.meta["example_value"]
+        result = hasattr(val, name)
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:134]: This doesn't look sound. val is a FakeTensor and in general its `__dict__` will not coincide with the actual tensor dict going on here. This should be modeled explicitly with an actual `__dict__` on Tensor. The irritating part is dealing with all of the "normal" attributes that show up on a regular Tensor, there are a lot of them. Class attributes I guess you can just test on torch.Tensor, but for the rest maybe we should be forced to handle them one by one.

File: torch/_dynamo/variables/tensor.py [hunk 7]
```diff
@@ -124,6 +128,14 @@ def check_type(ty):
...
+        return variables.ConstantVariable(result, **options).add_guard(
+            AttrSource(self.source, name).make_guard(GuardBuilder.HASATTR)
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:136]: You don't want to hasattr guard if you're just testing something like `hasattr(x, 'dim')` right? (Actually, now I'm also confused what this code does if there is no source on this VariableTracker)

**Pull request closed without merging**